usr/src/Xapps/xview3/clients/olwm/olwm.man
.\" @(#)olwm.man	1.1 10/4/91 SMI
.TH olwm 1 "6 August 1991"
.SH NAME
olwm \- OPEN LOOK window manager for OpenWindows
.SH SYNOPSIS
.B olwm
[
.I options
]
.\" ========================================================================
.SH DESCRIPTION
.LP
.B olwm
is a window manager for the X Window System that implements
parts of the OPEN LOOK graphical user interface.  It is the standard window
manager for Sun's OpenWindows product, but it will work properly with any
X11 system.
.\" ========================================================================
.SH OPTIONS
.LP
Most command-line options have counterparts in the resource database.  A 
command-line option will override any setting from the resource database.
.TP
.B \-2d
Use two-dimensional look.  This is the default for monochrome systems.
.TP
.B \-3d
Use three-dimensional look.  This is the default for color systems.  This
option is ignored for monochrome systems.
.TP
.BR \-bd ",\ \ " \-bordercolor
Specifies the border color.  See the description of the
.B BorderColor
resource.
.TP
.BR \-bg ",\ \ " \-background
Specifies the background color.  See the description of the
.B Background
resource.
.TP
.BR \-c ",\ \ " \-click
Use click-to-focus mode.  This is the default focus mode.
.TP
.BI "\-depth " depth
Specifies the depth of the visual in which
.B olwm
is to run.  See the discussion in the Screen Resources section.
.TP
.BI "\-display " display-string
Specify the name of the display to manage.  Overrides the DISPLAY
environment variable, if any.  In addition, the display string is exported
to
.BR olwm 's
environment, so processes forked from
.B olwm
will inherit this value.
.TP
.BR \-f ",\ \ " \-follow
Use focus-follows-mouse mode.  Default mode is click-to-focus.
.TP
\fB\-fn\fP\ \fIfont-name,\fP\ \ \fB\-font\fP\ \ \fIfont-name\fP
Set the font for window titles.
.TP
.BR \-fg ",\ \ " \-foreground
Specifies the foreground color.  See the description of the
.B ForegroundColor
resource.
.TP
.BI "\-multi"
Manage windows on all screens that a display supports.
This is the default.
.TP
.BI "\-name " resource-name
Use
.I resource-name
to look up resources in the resource database.
.TP
.BI "\-xrm " resource-string
Specify resources on the command-line.  Resources specified here will override
resources found in resource files.
.TP
.BI "\-single"
Manage windows for a single screen only, using the default screen
for the specified display.  Overrides the
.B \-multi
option.
.TP
.BI "\-syncpid " process-id
When
.B olwm
has completed its initialization, it will send a signal (SIGALRM by default)
to
.I process-id.
The signal will be sent only if this option is present.
This is useful for running
.B olwm
from shell scripts (such as .xinitrc) in such a way that the script waits
for
.B olwm
to finish its initialization, while leaving
.B olwm
as a child process of the shell script.
This can be done using the following
.BR sh (1)
construct:
.nf
.sp
	sleep 15 & pid=$!
	olwm \-syncpid $pid &
	wait $pid
.sp
.fi
.TP
.BI "\-syncsignal " signal
Specifies the signal to send instead of SIGALRM.  The signal is specified as
a number, not symbolically.
.TP
.BI "\-visual " visual-class
Specifies the class of the visual in which
.B olwm
is to run.  See the discussion in the Screen Resources section.
.\" ========================================================================
.SH DEBUGGING OPTIONS
.LP
The following options are strictly for debugging.  They are not recommended
for general use.  Don't use them
unless you know what you are doing.
.TP
.B \-all
Print a message for every event received.
.TP
.B \-debug
Equivalent to turning on all debugging options.
.TP
.B \-orphans
Print orphaned events.  Orphaned events are 
events that are associated with a window or frame that has no entry in
the frame hash table, or events that are not handled by the various
event handlers.
.TP
.B \-synchronize
Run the window manager in synchronous mode.
.\" ========================================================================
.SH INTERNATIONALIZATION OPTIONS
.TP
.BI "\-basiclocale " locale-name
Specify the basic OPEN LOOK locale category setting.
This category will be the base for other locale categories,
therefore there are certain restrictions applied to other locale
categories (see the following locale handling sections).
.TP
.BI "\-displaylang " locale-name
Specify the display language OPEN LOOK locale category.
This category affects the contents of workspace menu,
window menu and notice messages.
.TP
.BI "\-numeric " locale-name
Specify the numeric format OPEN LOOK locale category.
This category affects the numeric format displayed in
any message that contains numerics.
.\" ========================================================================
.SH LOCALE HANDLING
.PP
.I Locale
is the language and cultural conventions used in the program.
Locale will control the language-dependent part of
.BR olwm 's
behavior.
The OPEN LOOK international extensions have defined
several OPEN LOOK locale categories as follows:
.TP
.B "Basic Locale"
This is the basic setting for the entire locale mechanism.
This category specifies internal character handling behavior.
.TP
.B "Display Language"
This category specifies the language used for displaying menus,
notice messages and error messages.
.TP
.B "Input Language"
This category specifies the language used for text input.
This category has no affect on
.BR olwm ,
because it does not accept text input from the keyboard.
.TP
.B "Date Format"
This category specifies the format of date and time.
This category has no affect on
.BR olwm ,
because it does not display any date and time information.
.TP
.B "Numeric Format"
This category specifies the format of displayed numeric data.
.PP
Since the Basic Locale setting defines the character handling of
.BR olwm ,
certain restrictions have been placed on combining
the locale category settings.
.IP 1. 3
If basic setting is set to other than "C" locale,
then all other locale categories must be the same as 
basic setting or "C".
.IP 2.
If basic setting is set to "C" locale, then
all other locale categories must be in "C" locale.
.PP
The following methods are available to inform 
.BR olwm ,
the locale settings, and they are listed in order of priority:
.IP 1. 3
Command line options (such as
.BR \-basiclocale).
.IP 2.
by resource database.
.IP 3.
.BR setlocale (3)
function defaults (for example: LANG environment variable).
.PP
If command line options are not used to specify locale settings,
.B olwm
will pick up the new locale settings from
workspace property (per changes to the resource database);
and change the locale settings dynamically (for example, to change the
language of workspace and window menus).
All pinned workspace menus will be unpinned during this locale switching
operation.
.\" ========================================================================
.SH INPUT FOCUS
The
.I
input focus
.P
is the window that will receive keystrokes.
.B olwm
has two different input focus modes,
which are different ways of transferring the input focus from one
window to another.  By default,
.B olwm
uses "click-to-focus" (also known as "click-to-type") mode.  This means that
you must click on the window in order to get the focus to it.  While a
window has the input focus, the mouse can be anywhere on the screen; the
keyboard events will still go to that window.  You can set the input focus
to a window and simultaneously raise it to the top by clicking the left
mouse button in the window's title bar or border.
.LP
.B olwm
has another focus mode called "focus-follows-mouse."  In this mode, whatever
window the mouse is pointing to will receive the input focus.  To switch the
input focus from one window to another, you simply move the mouse to the
other window; you don't have to click at all.  Note, however, that to
transfer the focus amongst subwindows of a single top-level window, you must
click in the subwindow, or you must use focus transfer function keys (if
available from the application).
.LP
The input focus mode can be controlled with command-line options or by
entries in the resource database.  Neither focus mode has inherent
advantages.  Which one you choose is a matter of personal preference.
.\" ========================================================================
.SH MOUSE BUTTONS
OPEN LOOK defines three mouse button functions: SELECT, ADJUST, and MENU.
These functions are mapped to mouse buttons 1, 2, and 3, respectively.  On
systems with only two mouse buttons, the MENU function can be obtained by
pressing buttons 1 and 2 simultaneously.  This technique is referred to as
mouse button
.I chording.
The exact behavior of each of these functions depends on what object is
under the pointer.
.\" ========================================================================
.SH MANIPULATING WINDOWS AND ICONS
.I
Window Title Bar and Borders.
.P
Clicking SELECT selects the window, raises it above other windows, and
deselects any other objects.  In click-focus mode, the focus is also
transferred to this window.  Pressing and holding SELECT and then dragging
the mouse will move windows without raising them or setting the focus.  If
this window is selected, it and all other selected windows are moved
simultaneously.  Otherwise, just this window is moved, and it is not
selected.  If you hold down the Control key while you are moving a window,
motion is constrained to be either vertical or horizontal, depending on
whether you've moved farther in a vertical or horizontal direction.
Double-clicking SELECT on the window is the same as selecting the Full Size
(or Restore Size) menu item.  Clicking ADJUST will toggle the selected state
of this window.  If other windows or icons are already selected, they remain
selected.  ADJUST is useful for selecting several windows and icons.
Pressing MENU will bring up the window menu.  See the Window Menu section
for further details.  If the
.B Alt
key is held down, the mouse button functions become accessible anywhere over
the window, not just over the title bar and borders.  The modifier used can
be changed; see the description of the
.B WMGrab
resource in the section on Modifier Customization.
.LP
.I
Resize Corners.
.P
You can resize a window by pressing the left mouse button over any of the
resize corners and dragging it to the new location.  Releasing the mouse
button will set the new size of the window.  If you hold down the Control
key while you are dragging, the resize operation is contrained to resize
vertically or horizontally, depending on whether you've moved the mouse
farther in the horizontal or vertical direction.
.LP
.I
Window Button.
.P
The Window Button is the small box with a downward-pointing triangle near
the left end of the title bar.  Pressing MENU over the window button will
bring up the Window Menu.  Clicking SELECT over the left mouse button on the
Window Button will execute the window menu's default action.  This will
usually close the window into an icon.  You can change the window menu's
default action by holding down the Control key while manipulating the window
menu.
.LP
.I
Pushpin.
.P
OPEN LOOK pop-up windows have a pushpin instead of a window button.  If the
pushpin is out of its hole, pressing a command button within the window will
cause the window to be taken down ("dismissed") after the command is
executed.  If you click SELECT on the pushpin, it will move into its hole.
In this state, pressing a command button will execute the command without
dismissing the window.  Clicking SELECT over the pin will pull it out of the
hold.  This will dismiss the window without executing any commands.  Some
windows come up with the pin already in the hole.
.LP
.I
Icons.
.P
An icon represents a closed window.  You can still do most of the same
operations as with an open window.  Moving and selecting icons with SELECT
and ADJUST is exactly the same as for open windows.  A similar version of
the Window Menu is available on an icon by pressing MENU.  Double-clicking
SELECT will open the icon.  Icons cannot be resized.
.\" ========================================================================
.SH NONRECTANGULAR WINDOWS
The X11 Non-Rectangular Window Shape Extension (commonly referred to simply
as the Shape extension) allows windows to have arbitrary shapes.
.B Olwm
will handle these windows by giving them no decoration whatsoever.  Shaped
windows can be manipulated by using the WMGrab modifier (Alt by default)
with the mouse buttons.  (See the section on Modifier Customization for
further details.)  Shaped windows can be moved, resized, closed, opened, etc.
like ordinary windows.  The selection feedback for shaped windows is the
presence of resize corners floating at the corners of the bounding rectangle
of the window's shape.
.\" ========================================================================
.SH SELECTIONS ON THE WORKSPACE
.LP
You can select a group of windows and icons by using the left or middle
mouse buttons over the Workspace (the area of the screen outside of all
windows and icons, commonly known as the "root window").  Pressing either
SELECT or ADJUST and dragging the mouse will define a rubber-band rectangle.
When you release the mouse button, the set of windows and icons enclosed by
this rectangle will be operated on.  If you created the rectangle using
SELECT, the windows and icons within will be selected, and all other objects
will be deselected.  If you used ADJUST, the objects within will have their
selected state toggled, and any other windows and icons already selected
will remain selected.
.\" ========================================================================
.SH WORKSPACE MENU
.LP
Pressing MENU over the workspace brings up the Workspace Menu.  This menu is
customizable, but it typically contains at least the following items.
(The items may appear in a different language depending on the current
locale setting.)
.TP
.B Programs
This button has a sub-menu that allows you to invoke applications.  The
default Programs sub-menu contains all of the programs in the OpenWindows
DeskSet.  However, users typically customize this menu to contain many more
programs and to contain nested submenus.  See the section on Menu
Customization for further information.
.TP
.B Utilities
This button has a sub-menu that contains several utility functions for the
workspace, including Refresh (redisplay all windows on the screen), Lock
Screen, and Save Workspace.
.TP
.B Properties...
This item brings up the Workspace Properties window, which allows you to
view and customize settings of the OpenWindows environment.
.TP
.B Help...
Brings up the table of contents of the Help Handbooks.
.TP
.B "Desktop Intro..."
Brings up a tutorial introduction to the Sun Desktop.
.TP
.B Exit
Shuts down all applications and exits the window system.  A confirmation
notice is popped up first to give you a chance to cancel the operation.
.\" ========================================================================
.SH WINDOW MENU
.LP
The window menu of most windows has the following items.
(The items may appear in a different language depending on
the current locale setting.)
.TP
.B Close 
Close the window to an icon.  Any OPEN LOOK pop-up windows are closed into
this icon as well.  They will reappear when the icon is opened.  This item
is "Open" if you bring up the menu on an icon.
.TP
.B Full Size
Expand the window to the full height of the screen.  If this has already
done, the button is Normal Size instead of Full Size.  Normal Size restores
the window to the size it was before you did the Full Size operation.  If
the application has specified a maximum size for the window, this size is
used for Full Size instead of the full screen height.
.TP
.B Move
Starts the keyboard-based form of moving the window.  Appears only if OPEN
LOOK Mouseless Mode is enabled.
.TP
.B Resize
Starts the keyboard-based form of resizing the window.  Appears only if OPEN
LOOK Mouseless Mode is enabled.
.TP
.B Back
Move the window behind all other windows.
.TP
.B Refresh
Clear and redisplay the window.
.TP
.B Quit
Kill the program running in the window and remove the window.  If the
application has elected to participate in the WM_DELETE_WINDOW protocol,
.B olwm
sends a WM_DELETE_WINDOW ClientMessage instead of killing that window.
.LP
OPEN LOOK pop-up windows (as opposed to base windows) have a smaller window
menu.  It lacks the Close, Full Size, and Quit items, but it has two new
items:
.TP
.B Dismiss
Causes the window to be dismissed.  This button has a submenu with two
items: This Window, which dismisses just this window, and All Pop-ups, which
dismisses all pop-up windows owned by this application.
.TP

usr/src/Xapps/xview3/clients/olwm/olwm.c
/*
 *      (c) Copyright 1989, 1990 Sun Microsystems, Inc. Sun design patents
 *      pending in the U.S. and foreign countries. See LEGAL_NOTICE
 *      file for terms of the license.
 */

#ident	"@(#)olwm.c	26.42	91/09/14 SMI"

#include <errno.h>
#include <memory.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>

#include <sys/time.h>
#include <sys/types.h>

#include <sys/param.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/wait.h>

#include <X11/Xos.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/Xresource.h>

#include "i18n.h"
#include "ollocale.h"
#include "events.h"
#include "mem.h"
#include "olwm.h"
#include "win.h"
#include "menu.h"
#include "defaults.h"
#include "resources.h"
#include "globals.h"
#include "group.h"
#include "slots.h"

typedef	void	(*VoidFunc)();

/*
 * Globals
 */
char	*ProgramName;
char	*AppName;		/* strictly last component of ProgramName */
GlobalResourceVariables GRV;
XrmDatabase OlwmDB;


/*
 * Global Quarks.  "Top" refers to the root of the resource name/instance 
 * hierarchy.
 */
XrmQuark TopClassQ;
XrmQuark TopInstanceQ;
XrmQuark OpenWinQ;

/* Current Display */
Display	*DefDpy;


static Display *openDisplay();
static void	parseCommandline();
static void	sendSyncSignal();


/* Note that changes to RMTable instance names must also be made here !! */
static	XrmOptionDescRec	optionTable[] = {
	{ "-display",		".display",	
		XrmoptionSepArg, (caddr_t) NULL },
	{ "-f",			".setInput",
		XrmoptionNoArg, (caddr_t) "follow" },
	{ "-follow",		".setInput", 
		XrmoptionNoArg, (caddr_t) "follow" },
	{ "-c", 		".setInput",	
		XrmoptionNoArg, (caddr_t) "select" },
	{ "-click", 		".setInput",	
		XrmoptionNoArg, (caddr_t) "select" },
	{ "-2d", 		".use3D",	
		XrmoptionNoArg, (caddr_t) "False" },
	{ "-3d", 		".use3D",	
		XrmoptionNoArg, (caddr_t) "True" },
	{ "-parent", 		".reparentOK",	
		XrmoptionNoArg, (caddr_t) "False" },
	{ "-orphans", 		".printOrphans", 
		XrmoptionNoArg, (caddr_t) "True" },
	{ "-all", 		".printAll", 
		XrmoptionNoArg, (caddr_t) "True" },
	{ "-synchronize",	".synchronize",
		XrmoptionNoArg, (caddr_t) "True" },
	{ "-debug",	".printOrphans", XrmoptionNoArg, (caddr_t) "True" },
	{ "-name",		".name",	
		XrmoptionSepArg, (caddr_t) NULL },
	{ "-bg",	"*Background",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-background", "*Background",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-fg",	"*Foreground",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-foreground","*Foreground",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-bd",	"*Background",	
		XrmoptionSepArg, (caddr_t) NULL },
	{ "-bordercolor",	"*Background",	
		XrmoptionSepArg, (caddr_t) NULL },
	{ "-fn",	"*TitleFont",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-font",	"*TitleFont",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-xrm",	NULL,		XrmoptionResArg, (caddr_t) NULL },
	{ "-single",	".singleScreen",XrmoptionNoArg, (caddr_t) "True" },
	{ "-multi",	".singleScreen",XrmoptionNoArg, (caddr_t) "False" },
	{ "-syncpid",   ".syncPid",     XrmoptionSepArg, (caddr_t) NULL },
	{ "-syncsignal",".syncSignal",  XrmoptionSepArg, (caddr_t) NULL },
	{ "-depth",	"*depth",	XrmoptionSepArg, (caddr_t) NULL },
	{ "-visual",	"*visual",	XrmoptionSepArg, (caddr_t) NULL },
#ifdef OW_I18N_L3
        { "-basiclocale", "*basicLocaleCL", XrmoptionSepArg, (caddr_t) NULL },
       	{ "-displaylang", "*displayLangCL", XrmoptionSepArg, (caddr_t) NULL },
       	{ "-inputlang",   "*inputLangCL", XrmoptionSepArg, (caddr_t) NULL },
       	{ "-numeric",     "*numericCL", XrmoptionSepArg, (caddr_t) NULL },
       	{ "-dateFormat",  "*dateFormatCL", XrmoptionSepArg, (caddr_t) NULL },
#endif /* OW_I18N_L3 */
};
#define OPTION_TABLE_ENTRIES (sizeof(optionTable)/sizeof(XrmOptionDescRec))

static void	initWinClasses();

/* Child Process Handling */
void HandleChildSignal();
void ReapChildren();
int DeadChildren = 0;
int SlavePid;

#ifdef ALLPLANES
int allplanes;			/* server supports the ALLPLANES extension */
#endif

#ifdef SHAPE
Bool ShapeSupported;		/* server supports the SHAPE extension */
int  ShapeEventBase;
int  ShapeErrorBase;
#endif


static char	**argVec;

/*
 * main	-- parse arguments, perform initialization, call event-loop
 *
 * The OpenLook window manager was written before the OpenLook
 * spec. was completed.  Because of this, some of it works in
 * an awkward way.  Specifically, the way a window's focus and
 * select state is shown.
 *
 * For example, when a window is focused in click-to-focus mode
 * the header gets highlighted, (black background, white text).
 * As it was written, the titlebar window sits just inside of the 
 * resize corners.  Because the OpenLook spec. requires the header to 
 * be highlighted in line with the resize corners, blacking in the
 * the background of the titlebar window is one pixel short in each
 * direction of being a big enough highlighted area.  We can't make
 * the titlebar bigger because it would then overlap the resize corners.
 * Now that the spec. is complete, OLWM should be restructured.
 *
 * Overview:
 *
 *	Upon startup OLWM reparents all the client windows unless
 *	the user has told it otherwise (-p option).
 * 	OLWM creates a frame which will contain the client window (the pane)
 *  	and the decoration windows (resize corners, titlebar).  The
 *	titlebar window contains the pushpin, if any.
 *
 *	After creating the decorations for the windows OLWM sits
 * 	in a forever loop getting the next event and calling the
 * 	corresponding window's event handling routine.
 *
 *	Each window has associated with it an event handling 
 * 	routine (EventFrame, EventPushPin) which the main EventLoop
 *	will call when an event happens in that window.  The event
 *	handling routine will call the proper routines to move a 
 *	window, create a menu, etc.
 */
main(argc, argv)
	int argc;
	char **argv;
{
	int			ErrorHandler();
	int			ExitOLWM();
	XrmDatabase		commandlineDB = NULL;
	char			*dpystr;

#ifdef OW_I18N_L3
#ifndef MAXPATHLEN
#define MAXPATHLEN 256
#endif
	char			*OpenWinHome;
	char			locale_dir[MAXPATHLEN+1];
	extern char		*getenv();
#endif /* OW_I18N_L3 */

#ifdef MALLOCDEBUG
	malloc_debug(MALLOCDEBUG);
#endif /* MALLOCDEBUG */
#ifdef GPROF_HOOKS
	moncontrol(0);
#endif /* GPROF_HOOKS */

#ifdef OW_I18N_L3
       	/*
       	 * Even in the SUNDAE1.0 (first release) we might need the
       	 * dynamic locale change for window manager, since window
       	 * manager is usually never re-start again in one sesstion.
       	 * But leave static for now.
       	 */
       	/*
       	 * We are setting the locale (issuing the setlocale) by
       	 * EffectOLLC() function, but we need to call setlocale here
       	 * to handle command line argument with certain locale.
       	 * FIX_ME! But may not work well, because we did not touch the
       	 * Xlib function XrmParseCommand().
       	 */
       	if (setlocale(LC_ALL, "") == NULL) {
		/*
		 * FIX_ME: Following message may be misleading,
		 * because setlocale will retrieve more than just a
		 * $LANG environment variable (such as $LC_CTYPE).
		 * Also, later if the resource DB has a locale
		 * infomation, we will use it. Hence, this setting is
		 * may be only for the command line processing.
		 */
		fprintf(stderr, gettext("\
		olwm: Warning: '%s' is invalid locale, using 'C' locale.\n"),
				getenv("LANG"));
		(void)setlocale(LC_ALL,"C");
	}
	if ((OpenWinHome = getenv("OPENWINHOME")) != 0)
		(void)strcpy(locale_dir,OpenWinHome);
	else
		(void)strcpy(locale_dir,"/usr");
	(void)strcat(locale_dir,"/lib/locale");
	bindtextdomain("olwm_messages",locale_dir);
       	textdomain("olwm_messages");
#endif /* OW_I18N_L3 */

	ProgramName = argv[0];
	argVec = argv;

	/* Set up a signal handler so that if we are killed we put
	 * the windows back into a usable state.  Specifically we
	 * need to reset the focus so they will be able to get input.
	 */
	signal(SIGHUP, (VoidFunc)ExitOLWM);
	signal(SIGINT, (VoidFunc)ExitOLWM);
	signal(SIGTERM, (VoidFunc)ExitOLWM);
	signal(SIGCHLD, HandleChildSignal);

	/* initialize the resource manager */
	XrmInitialize();

	/* parse the command line arguments into local tmp DB */
	parseCommandline( &argc, argv, &commandlineDB );

	DefDpy = openDisplay(commandlineDB);

#ifdef ALLPLANES
	{
	    int tmp;
	    allplanes = XAllPlanesQueryExtension(DefDpy, &tmp, &tmp);
	}
#endif /* ALLPLANES */


#ifdef SHAPE
	ShapeSupported = XShapeQueryExtension(DefDpy, &ShapeEventBase,
					      &ShapeErrorBase);
#endif /* SHAPE */

	/* put all resources into global OlwmDB and set olwm variables */
	GetDefaults(DefDpy, commandlineDB);

	/* Initialize the event handling system. */
	InitEvents(DefDpy);
	InitBindings(DefDpy);

	/* Set up the error handling function. */
	XSetErrorHandler(ErrorHandler);

	/* this can be useful for debuging strange client behaivior */
	if (GRV.Synchronize)
		XSynchronize(DefDpy, True);

	/* Initialize the atoms we will need to communicate with
	 * Open Look clients.
	 */
	InitAtoms(DefDpy);

	/* Initialize the database. */
	WIInit( DefDpy );

	/* Initialise all types of window classes */
	initWinClasses(DefDpy);

	/* Initialize client led state */
	InitClientState(DefDpy);

	/* initialise the group module */
	GroupInit();

	/*
	 * Force child processes to disinherit the TCP file descriptor.
	 * This helps shell commands forked and exec'ed from menus
	 * to work properly.
	 */
	if (fcntl(ConnectionNumber(DefDpy), F_SETFD, 1) == -1)
	{
		perror( gettext("olwm: child cannot disinherit TCP fd") );
		exit( -1 );
	}

#ifdef OW_I18N_L3
	EffectOLLC(DefDpy);
#endif /* OW_I18N_L3 */
	/* Init the global menus */
	InitMenus(DefDpy);

	/* init region handling code */
	InitRegions();

	/* Init screen */
	InitScreens(DefDpy);
	GrabKeys(DefDpy, True);
	GrabButtons(DefDpy, True);
	ReparentScreens(DefDpy);
	if (!GRV.FocusFollowsMouse)
	    ClientFocusTopmost(DefDpy, GetFirstScrInfo(), CurrentTime);

	/*
 	 * Start olwmslave - using the same args we got.
	 */
	if (GRV.RunSlaveProcess)
	    SlavePid = SlaveStart(argVec);

	if (GRV.Beep != BeepNever)
	    XBell(DefDpy, 100);

	sendSyncSignal();

	EventLoop( DefDpy );

	/*NOTREACHED*/
}


/* 
 * parseCommandline - parse the command line arguments into a resource
 *	database
 */
static void
parseCommandline( argc, argv, tmpDB )
int		*argc;
char		*argv[];
XrmDatabase	*tmpDB;
{
	char	instName[MAX_NAME];
	char	namestr[MAX_NAME];
	char	*type, *p;
	XrmValue val;

	/* Extract trailing pathname component of argv[0] into AppName. */

	AppName = strrchr(argv[0], '/');
	if (AppName == NULL)
	    AppName = argv[0];
	else
	    ++AppName;

	XrmParseCommand(tmpDB, optionTable, OPTION_TABLE_ENTRIES,
			AppName, argc, argv );

	/*
	 * Initialize root instance and class quarks.  Create the instance
	 * name by first looking up the "name" resource in the command line
	 * database (for the -name option).  If it's not present, use AppName
	 * (the trailing pathname component of argv[0]).  Then, scan it and
	 * replace all illegal characters with underscores.  Note: we don't
	 * use the ctype functions here, because they are internationalized.
	 * In some locales, isalpha() will return true for characters that are
	 * not valid in resource component names.  Thus, we must fall back to
	 * standard character comparisions.
	 *
	 * REMIND: specifying the -name option changes the name with which 
	 * resources are looked up.  But the command line options were put 
	 * into the database using AppName, which is based on argv[0].  Thus, 
	 * specifying -name causes all command-line args to be ignored, which 
	 * is wrong.
	 */

	(void) strcpy(namestr, AppName);
	(void) strcat(namestr, ".name");
	if (XrmGetResource(*tmpDB, namestr, namestr, &type, &val)) {
	    (void) strncpy(instName, (char *)val.addr, MAX_NAME);
	} else {
	    (void) strncpy(instName, AppName, MAX_NAME);
	}

	instName[MAX_NAME-1] = '\0';
	for (p=instName; *p != '\0'; ++p) {
	    if (!(*p >= 'a' && *p <= 'z' ||
		  *p >= 'A' && *p <= 'Z' ||
		  *p >= '0' && *p <= '9' ||
		  *p == '_' || *p == '-')) {
		*p = '_';
	    }
	}
	TopInstanceQ = XrmStringToQuark(instName);
	TopClassQ = XrmStringToQuark("Olwm");
	OpenWinQ = XrmStringToQuark("OpenWindows");

	/* check to see if there are any arguments left unparsed */
	if ( *argc != 1 )
	{
		/* check to see if it's -help */
		if ( argv[1][0] == '-' && argv[1][1] == 'h' )
			usage(gettext("Command line arguments accepted"),gettext("are:"));
		else
			usage(gettext("Unknown argument(s)"), gettext("encountered"));
	}
}


/*
 * openDisplay
 */
static Display *
openDisplay(rdb)
    XrmDatabase rdb;
{
    char namebuf[MAX_NAME];
    char *type;
    XrmValue value;
    char *dpystr = NULL;
    char *envstr;
    Display *dpy;

    (void) strcpy(namebuf, AppName);
    (void) strcat(namebuf, ".display");

    if (XrmGetResource(rdb, namebuf, namebuf, &type, &value)) {
	dpystr = (char *)value.addr;
	envstr = (char *)MemAlloc(8+strlen(dpystr)+1);
	sprintf(envstr, "DISPLAY=%s", dpystr);
	putenv(envstr);
    }

    dpy = XOpenDisplay(dpystr);
    if (dpy == NULL) {
	fprintf(stderr, gettext("%s: cannot connect to %s\n"),
		ProgramName, dpystr);
	exit(1);
    }
    return dpy;
}


/*
 * sendSyncSignal
 *
 * Send a signal to the process named on the command line (if any).  Values
 * for the process id and signal to send are looked up in the resource 
 * database; they are settable with command-line options.  The resources are 
 * looked up with the names
 * 
 *	<appname>.syncPid		process id
 *	<appname>.syncSignal		signal to send (integer)
 *
 * where <appname> is the trailing pathname component of argv[0].
 */
static void
sendSyncSignal()
{
    char *type;
    XrmValue value;
    int pid;
    int sig = SIGALRM;
    int tmp;
    char namebuf[100];

    (void) strcpy(namebuf, AppName);
    (void) strcat(namebuf, ".syncPid");
    if (!XrmGetResource(OlwmDB, namebuf, namebuf, &type, &value))
	return;
    pid = atoi((char *)value.addr);
    if (pid <= 0 || pid > MAXPID)
	return;

    (void) strcpy(namebuf, AppName);
    (void) strcat(namebuf, ".syncSignal");
    if (XrmGetResource(OlwmDB, namebuf, namebuf, &type, &value)) {
	tmp = atoi((char *)value.addr);
	if (tmp > 0 && tmp <= SIGUSR2)
	    sig = tmp;
    }
    (void) kill(pid, sig);
}
 

static void
initWinClasses(dpy)
Display *dpy;
{
	FrameInit(dpy);
	IconInit(dpy);
	ResizeInit(dpy);
	ColormapInit(dpy);
	ButtonInit(dpy);
	BusyInit(dpy);
	MenuInit(dpy);
	PinMenuInit(dpy);
	RootInit(dpy);
	NoFocusInit(dpy);
	PushPinInit(dpy);
	PaneInit(dpy);
	IconPaneInit(dpy);
}


/*
 * Exit -- exit routine called by menus.
 *	Kill all running applications.
 */
Exit(dpy)
Display	*dpy;
{
	extern void *ClientShutdown();
	
	SlaveStop();
	ListApply(ActiveClientList, ClientShutdown, (void *)0);
	XSync(dpy, True);
	exit(0);
	/*NOTREACHED*/
}


static void
cleanup()
{
	extern void *UnparentClient();

	/*
 	 * If DefDpy is NULL then we didn't get to the XOpenDisplay()
	 * so basically there is nothing to clean up so return.
	 */
	if (DefDpy == NULL)
		return;

	/*
	 * Stop olwmslave
 	 */
	SlaveStop();

	/*
	 * destroy all pinned menus
	 */
	DestroyPinnedMenuClients();

	/*
	 * Go through the list of windows.  Unmap all icons that are on the
	 * screen.  Reparent all windows back to the root, suitably offset
	 * according to their window-gravities.  Also remap all non-withdrawn
	 * windows, and remove all Withdrawn windows from the save-set (so
	 * they don't get remapped.  REMIND: We have to do this because
	 * Withdrawn windows are still left reparented inside the frame; this
	 * shouldn't be the case.
	 */
	ListApply(ActiveClientList,UnparentClient,NULL);

	/* Destroy the screens - which will restore input focus, colormap,
	 * and background, etc.
	 */
	DestroyScreens(DefDpy);

	XSync(DefDpy, True);
}


/* Clean up and then re-exec argv. */
int
RestartOLWM()
{
    cleanup();
    execvp(argVec[0], argVec);
    ErrorGeneral("cannot restart");
    /*NOTREACHED*/
}


/* Clean up and then exit. */
int
ExitOLWM()
{
    cleanup();
    exit(0);
}


/*
 * HandleChildSignal - keep track of children that have died
 */
static void
HandleChildSignal()
{
	++DeadChildren;
}

/*
 * ReapChildren - wait() for all dead child processes
 */
void
ReapChildren()
{
#ifdef SYSV
        sigset_t newmask, oldmask;
        int status;
        pid_t pid;
#else
	int oldmask;
	int pid;
	union wait status;
#endif

#ifdef SYSV
        sigemptyset(&newmask);
        sigemptyset(&oldmask);
        sigaddset(&newmask, SIGCHLD);
        sigprocmask(SIG_BLOCK, &newmask, &oldmask);
#else
	oldmask = sigblock(sigmask(SIGCHLD));
#endif
	while (DeadChildren>0)
	{
#ifdef SYSV
                pid = waitpid(-1, &status, WNOHANG);
#else
                pid = wait3(&status, WNOHANG, (struct rusage *)0);
#endif
		/* if it's the slave process then stop its use */
		if (pid == SlavePid) 
			SlaveStopped();

		if (pid == 0)
		{
			/* Removed error message - deemed not useful */
			DeadChildren = 0;
			break;
		}
		if (WIFSTOPPED(status))
		{
			kill(pid, SIGKILL);
		}
		--DeadChildren;
	}
#ifdef SYSV
        sigprocmask(SIG_SETMASK, &oldmask, &oldmask);
#else
        (void) sigsetmask(oldmask);
#endif
}

/*
 * usage(s1, s2)	-- print informative message regarding usage
 */
static
usage(s1, s2)
char	*s1, *s2;
{
    fprintf(stderr, "%s %s\n", s1, s2);
    fprintf(stderr,
	"usage: %s [-2d] [-3d] [-click] [-follow] [-parent]\n", ProgramName);
    fputs(
       /* STRING_EXTRACTION - do not translate "-display" ,"-name"
        * and "-xrm", becuase those are namne of command line option.
        */
	gettext("\t[-display <display>] [-name <classname>] [-xrm <rsrcstring>]\n"),
	stderr);
    fputs("\t[-all] [-debug] [-orphans] [-synchronize] [-single]\n", stderr);
    fputs("\t[-syncpid pid] [-syncsignal signal]\n", stderr);
    exit(1);
}

